lib/elixir/test/elixir/code_test.exs

```diff
@@ -153,6 +153,36 @@ defmodule CodeTest do
                {:error, {1, "unsafe atom does not exist: ", "there_is_no_such_atom"}}
     end
 
+    test "supports static_atoms_encoder" do
+      ref = make_ref()
+
+      encoder = fn "there_is_no_such_atom", metadata ->
+        assert metadata[:line] == 1
+        assert metadata[:column] == 1
+        assert metadata[:file] == "nofile"
+        {:ok, {:my, "atom", ref}}
+      end
+
+      assert {:ok, {:my, "atom", ^ref}} =
+               Code.string_to_quoted(":there_is_no_such_atom", static_atoms_encoder: encoder)
+    end
+
+    test "extended static_atoms_encoder" do
+      encoder = fn string, _metadata ->
+        try do
+          {:ok, String.to_existing_atom(string)}
+        rescue
+          ArgumentError ->
+            {:ok, {:user_atom, string}}
+        end
+      end
+
+      assert {:ok, {:try, _, [[do: {:test, _, [{{:user_atom, "atom_does_not_exist"}, _, []}]}]]}} =
+               Code.string_to_quoted("try do: test(atom_does_not_exist())",
+                 static_atoms_encoder: encoder
+               )
+    end
```

Comment:
We also need a test for the `{:error, reason}` branch. For instance, I think we will have to convert the Reason to a charlist and the current code does not do it!

Reply:
Yep you are right. Sorry I forgot, will add this. The return value for the error was not entirely clear for me (the spec just says `{:error, {line :: pos_integer, term, term}}`).

Reply:
Fix this in 282675e0570a1e0f6e7be0579f7495e11ba4bf9d

Note that it adds `": "` to the error message. Which is kind of weird but it is in line with the other error messages. Alternatively it could not do that but then the encoder callback function is supposed to add it, which is also questionable..

Reply:
:+1: